providers/kubeconfig/provider.go

```diff
@@ -0,0 +1,593 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	mcmanager "sigs.k8s.io/multicluster-runtime/pkg/manager"
+	"sigs.k8s.io/multicluster-runtime/pkg/multicluster"
+)
+
+const (
+	// DefaultKubeconfigSecretLabel is the default label key to identify kubeconfig secrets
+	DefaultKubeconfigSecretLabel = "sigs.k8s.io/multicluster-runtime-kubeconfig"
+
+	// DefaultKubeconfigSecretKey is the default key in the secret data that contains the kubeconfig
+	DefaultKubeconfigSecretKey = "kubeconfig"
+)
+
+var _ multicluster.Provider = &Provider{}
+
+// New creates a new Kubeconfig Provider.
+func New(opts Options) *Provider {
+	// Set defaults
+	if opts.KubeconfigLabel == "" {
+		opts.KubeconfigLabel = DefaultKubeconfigSecretLabel
+	}
+	if opts.KubeconfigKey == "" {
+		opts.KubeconfigKey = DefaultKubeconfigSecretKey
+	}
+	if opts.ConnectionTimeout == 0 {
+		opts.ConnectionTimeout = 10 * time.Second
+	}
+	if opts.CacheSyncTimeout == 0 {
+		opts.CacheSyncTimeout = 30 * time.Second
+	}
+	if opts.KubeconfigPath == "" {
+		opts.KubeconfigPath = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+	}
+
+	return &Provider{
+		opts:        opts,
+		log:         log.Log.WithName("kubeconfig-provider"),
+		client:      nil, // Will be set in Run
+		clusters:    map[string]cluster.Cluster{},
+		cancelFns:   map[string]context.CancelFunc{},
+		seenHashes:  map[string]string{},
+		readySignal: make(chan struct{}),
+	}
+}
+
+// Options are the options for the Kubeconfig Provider.
+type Options struct {
+	// Namespace to watch for kubeconfig secrets
+	Namespace string
+
+	// Label key to identify kubeconfig secrets
+	KubeconfigLabel string
+
+	// Key in the secret data that contains the kubeconfig
+	KubeconfigKey string
+
+	// ConnectionTimeout is the timeout for connecting to a cluster
+	ConnectionTimeout time.Duration
+
+	// CacheSyncTimeout is the timeout for waiting for the cache to sync
+	CacheSyncTimeout time.Duration
+
+	// KubeconfigPath is the path to the kubeconfig file to use for development/testing
+	// If not set, will use the default ~/.kube/config
+	KubeconfigPath string
+}
+
+type index struct {
+	object       client.Object
+	field        string
+	extractValue client.IndexerFunc
+}
+
+// Provider is a cluster provider that watches for secrets containing kubeconfig data
+// and engages clusters based on those kubeconfigs.
+type Provider struct {
+	opts        Options
+	log         logr.Logger
+	client      client.Client
+	lock        sync.RWMutex
+	clusters    map[string]cluster.Cluster
+	cancelFns   map[string]context.CancelFunc
+	indexers    []index
+	seenHashes  map[string]string // tracks resource versions
+	readySignal chan struct{}     // Signal when provider is ready to start
+	readyOnce   sync.Once         // Ensure we only signal once
+}
```

Comment:
It's unclear what this lock protects. Usually one uses field groups to make it clear:

```suggestion
type Provider struct {
	opts        Options
	log         logr.Logger
	client      client.Client

	readySignal chan struct{}     // Signal when provider is ready to start
	readyOnce   sync.Once         // Ensure we only signal once

	lock        sync.RWMutex // protects everything below.
	clusters    map[string]cluster.Cluster
	cancelFns   map[string]context.CancelFunc
	indexers    []index
	seenHashes  map[string]string // tracks resource versions	
}
```

Reply:
> `// tracks resource versions`

go doc what the key and value is.

Reply:
Cool, I've simplified it to

```go
type Provider struct {
	opts      Options
	log       logr.Logger
	client    client.Client
	lock      sync.RWMutex // protects everything below.
	clusters  map[string]cluster.Cluster
	cancelFns map[string]context.CancelFunc
	indexers  []index
}
```